### PR TITLE
[NEW] Feedback survey monitor

### LINKED ISSUE #1384: [DF] Design new timer which sends an event to CX after 15 mins idle time.
Use case:

If the visitor is idle for 15 minutes, we would like to trigger CX to send a feedback survey. 
- This would be like Visitor Abandonment since we would have to reset the timer every time the visitor sends a message.
- After 15 mins of inactivity we would send an event to CX ("trigger_feedback" maybe) which would trigger the feedback survey.
- We might also need to reset the VA timer upon sending this new event, so that the visitor will have another 19 mins to answer the survey before the session times out. 

NOTE: This new 'trigger_feedback' event will only be required on some sessions.  The plan would be to only offer a visitor the feedback survey every N number of days (30 or 45 days).  
- Upon initial registration with CX for a new session, CX will send a new flag in the response ('trigger_feedback': true) and if this is present, we would set and track the feedback timer for this session. For all other sessions this would be a NOOP.

### PR DESCRIPTION
**Pull request by bhardwajaditya**
Closes: https://github.com/WideChat/Rocket.Chat/issues/1384

## Proposed changes (including videos or screenshots)

Enable this from settings->omnichannel-> Feedback Survey

image

The Dialogflow incoming URL can be copied from the following apge
image

## Steps to test or reproduce

Sample payload from CX

To enable survey monitor (Time needs to be in seconds)
```
{
  "action": {
    "params": {
      "eventName": "feedbackSurvey",
      "time": 60,
      "value": true
    },
    "name": "df_feedback_survey"
  }
}
```

When the survey is sent to the user
```
{
  "action": {
    "params": {
      "time": 60,
      "eventName": "feedbackSurvey",
      "value": false
    },
    "name": "df_feedback_survey"
  }
}
```

**The event name in cx needs to be feedbackSurvey**

### CODE REVIEW

File: apps/meteor/app/livechat/server/lib/FeedbackSurveyMonitor.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+			return;
+		}
+
+		const rooms = await LivechatRooms.findOpenWithFeedbackSurveyCustomField();
+		rooms.forEach(async (room: any) => {
+			const surveyTimer = room.customFields?.feedbackSurvey;
+			if (surveyTimer && room.lastMessage && room.lastMessage?.u?._id !== room?.v?._id) {
+				const currentTime = new Date();
+				const lastMessageTime = new Date(room.lastMessage.ts);
+				if (lastMessageTime.getTime() + surveyTimer * 1000 <= currentTime.getTime()) {
```
Comment (ear-dev) [inline apps/meteor/app/livechat/server/lib/FeedbackSurveyMonitor.ts:67]: @bhardwajaditya can we a log point for when we are triggering the feedback survey event for a particular room please?  Need to include the RoomId.  Thanks.

File: apps/meteor/packages/rocketchat-i18n/i18n/en.i18n.json [hunk 2]
```diff
@@ -1753,6 +1754,8 @@
   "Enable_CSP": "Enable Content-Security-Policy",
   "Enable_CSP_Description": "Do not disable this option unless you have a custom build and are having problems due to inline-scripts",
   "Enable_Desktop_Notifications": "Enable Desktop Notifications",
+  "Enable_Feedback_Survey": "Enable Feedback Survey",
+  "Enable_Feedback_Survey_Description": "Enable the feedback survey to be shown to users",
```
Comment (ear-dev) [inline apps/meteor/packages/rocketchat-i18n/i18n/en.i18n.json]: Can we change this description please:
"Enable the feedback survey monitor to support a Dialogflow custom payload indicating a user should be presented with the survey after inactivity."

I know that's pedantic....  but it's also kind of self documenting, which I think helpful in the long run.

**Comment by ear-dev** [comment]
@bhardwajaditya so far I have not been able to get the feedback survey monitor to work... it does not seem to be triggering the event.  I think those log points I requested will help me to see what's going on?

I also see a bug in the UI maybe?  Even after I set the feedback survey to 'enabled' and paste the incoming endpoint, when I refresh the page it is showing as disabled.  Any thoughts?

image

**Comment by ear-dev** [comment]
Oh wow......  now it's working......  :-)  Still need to fix the UI please...

**Comment by bhardwajaditya** [comment]
@ear-dev this settings UI work for
me, will see if I can
Reproduce I tested this. That log point for started and finished indicates if it's enabled or not. That's by the cron job in rc server. I didn't add that one

**Comment by ear-dev** [comment]
So this is working great now!  Thanks.  If you could add some log points for when a session is having the feedbackSurvey custom field set and then unset, I still think that would be helpful.  I'll keep looking at my UI issue, but thanks for verifying.

**Comment by ear-dev** [comment]
@bhardwajaditya I tested the UI on another browser and it seems to be fine.  Must be something in cache....

**Comment by bhardwajaditya** [comment]
@ear-dev for log points you want console log or SNS logs?

**Comment by ear-dev** [comment]
> @ear-dev for log points you want console log or SNS logs?

just console log please.

**Comment by bhardwajaditya** [comment]
@ear-dev log points added in both the PRs

**Review by ear-dev — APPROVED** [review]
This looks good!  Thanks!

**Pull request merged**